File: data/src/main/java/com/crhonvas/data/mappers/posts/PostsApiToDomainMapper.java
package com.crhonvas.data.mappers.posts;


import android.support.annotation.NonNull;

import com.crhonvas.data.remote.model.posts.PostsApiResponse;
import com.crhonvas.data.utils.ListUtils;
import com.crhonvas.domain.model.posts.Post;

import java.util.List;

import javax.inject.Inject;

//TODO: Inject scheduler provider inorder to do mappings in a specified background thread

/**
 * Mapper from {@link PostsApiResponse} to {@link Post}
 */
public class PostsApiToDomainMapper {

    @Inject
    public PostsApiToDomainMapper() {
    }

    @NonNull
    public Post map(@NonNull PostsApiResponse postsApiResponse) {
        if (postsApiResponse == null) {
            return new Post.Builder().build();
        }
        return new Post.Builder()
                .userId(postsApiResponse.getUserId())
                .id(postsApiResponse.getId())
                .title(postsApiResponse.getTitle())
                .body(postsApiResponse.getBody())
                .build();
    }

    @NonNull
    public List<Post> map(@NonNull List<PostsApiResponse> postsApiResponse) {
        return ListUtils.map(postsApiResponse, this::map);
    }
}

File: data/src/main/java/com/crhonvas/data/mappers/posts/PostsApiToDomainMapper.kt
package com.crhonvas.data.mappers.posts

import com.crhonvas.data.remote.model.posts.PostsApiResponse
import com.crhonvas.domain.model.posts.Post
import javax.inject.Inject

//TODO: Inject scheduler provider inorder to do mappings in a specified background thread
/**
 * Mapper from [PostsApiResponse] to [Post]
 */
class PostsApiToDomainMapper @Inject constructor() {
    fun map(postsApiResponse: PostsApiResponse): Post {
        return if (postsApiResponse == null) {
            Post.Builder().build()
        } else Post.Builder()
            .userId(postsApiResponse.userId)
            .id(postsApiResponse.id)
            .title(postsApiResponse.title)
            .body(postsApiResponse.body)
            .build()
    }

    fun map(postsApiResponse: List<PostsApiResponse>): List<Post> {
        return postsApiResponse.map { map(it) }
    }
}